File: patternlets/java-OpenMPI/14.reduction2/Reduction2.java
/* Reduction2.java
 * ... illustrates the reduction pattern on multiple values,
 *      using buffers in OpenMPI's Java interface.
 *
 * Joel Adams, Calvin University, November 2019.
 *
 * Usage: mpirun -np 4 java ./Reduction2
 *
 * Exercise:
 * - Compile, then run with N = 1, 2, 3, 4, 
 *     comparing output to source code.
 * - Explain behavior of reduce() in terms of
 *     srcBuf and destBuf.
 */

import mpi.*;
import java.nio.IntBuffer;

public class Reduction2 {

 public static void main(String [] args) throws MPIException {
    MPI.Init(args);

    Comm comm        = MPI.COMM_WORLD;
    int id           = comm.getRank();
    int numProcesses = comm.getSize();

    IntBuffer srcBuf = MPI.newIntBuffer(BUFFER_SIZE);
    IntBuffer destBuf = MPI.newIntBuffer(BUFFER_SIZE);

    if (id == MASTER) {
        System.out.print("\nBefore reduction: ");
        printBuf(id, "destBuf", destBuf); 
    }

    for (int i = 0; i < BUFFER_SIZE; ++i) {
        srcBuf.put(i, id * i);
    }

    printSeparator("", id);
    printBuf(id, "srcBuf", srcBuf);
    printSeparator("----", id);

    comm.reduce(srcBuf, destBuf, BUFFER_SIZE,
                 MPI.INT, MPI.SUM, MASTER);

    if ( id == MASTER) {
        System.out.print("After reduction: ");
        printBuf(id, "destBuf", destBuf);
        System.out.println();
    }

    MPI.Finalize();
  }

  /* utility to display the contents of an IntBuffer.
   * @param: id, the int MPI rank of this process.
   * @param: bufName, a String that is the name of the buffer.
   * @param: buf, the IntBuffer.
   * @param: size, the size of buf.
   */
  private static void printBuf(int id, String bufName, IntBuffer buf) {
      String msg = "Process " + id + ", " + bufName + ": [";
      int size = buf.capacity();
      int sizeLessOne = size - 1;
      for (int i = 0; i < size; ++i) {
         msg += buf.get(i);
         if (i < sizeLessOne ) {
             msg += ",";
         }
      }
      msg += "]\n";
      System.out.print(msg);
  }

  /* utility to print a separator between before and after sections.
   * @param: separator, a String.
   * @param: id, the MPI rank of this process. 
   * POST: separator has been printed by the master process.
   */
  private static void printSeparator(String separator, int id) throws MPIException {
     MPI.COMM_WORLD.barrier();
     if (id == MASTER) { System.out.println(separator); }
     MPI.COMM_WORLD.barrier();
  }

  private static int BUFFER_SIZE = 5;
  private static int MASTER      = 0;
}

File: patternlets/MPI/13.reduction2/reduction2.c
/* reduction2.c
 * ... illustrates the use of MPI_Reduce() using arrays...
 * Joel Adams, Calvin College, January 2015.
 *
 * Usage: mpirun -np 4 ./reduction2
 *
 * Exercise:
 * - Compile and run, comparing output to source code.
 * - Explain behavior of MPI_Reduce() in terms of
 *     srcArr and destArr.
 */

#include <mpi.h>
#include <stdio.h>

#define ARRAY_SIZE 5

void printArray(int id, char* arrayName, int* array, int SIZE);
void printSeparator(char* separator, int id);

int main(int argc, char** argv) {
    int myRank = -1;
    int srcArr[ARRAY_SIZE] = {0};
    int destArr[ARRAY_SIZE] = {0};

    MPI_Init(&argc, &argv);
    MPI_Comm_rank(MPI_COMM_WORLD, &myRank);

    if (myRank == 0) {
        printf("\nBefore reduction: ");
        printArray(myRank, "destArr", destArr, ARRAY_SIZE);
    }

    for (unsigned i = 0; i < ARRAY_SIZE; i++) {
        srcArr[i] = myRank * i;
    }

    printSeparator("", myRank);
    printArray(myRank, "srcArr", srcArr, ARRAY_SIZE);
    printSeparator("----", myRank);

    MPI_Reduce(srcArr, destArr, ARRAY_SIZE, MPI_INT, MPI_SUM, 0, MPI_COMM_WORLD);

    if (myRank == 0) {
        printf("\nAfter reduction:  ");
        printArray(myRank, "destArr", destArr, ARRAY_SIZE);
        printf("\n");
    }

    MPI_Finalize();

    return 0;
}

/* utility to display an array
 * params: id, the rank of the current process
 *         arrayName, the name of the array being displayed
 *         array, the array being displayed
 *         SIZE, the number of items in array.
 * postcondition:
 *         the id, name, and items in array have been printed to stdout.
 */
void printArray(int id, char* arrayName, int * array, int SIZE) {
    printf("Process %d, %s: [", id, arrayName);
    for (int i = 0; i < SIZE; i++) {
        printf("%3d", array[i]);
        if (i < SIZE-1) printf(",");
    }
    printf("]\n");
}

/* utility to print a separator string between before and after sections.
 * params: separator, a string
 *         id, the rank of the current process.
 * postcondition: the master process has printed the separator.
 */
void printSeparator(char* separator, int id) {
    MPI_Barrier(MPI_COMM_WORLD);
    if (id == 0) { printf("%s", separator); }
    MPI_Barrier(MPI_COMM_WORLD);
}


